pymc/gp/hsgp.py

```diff
@@ -0,0 +1,239 @@
+        if not isinstance(m, Sequence):
+            raise ValueError(arg_err_msg)
+        if len(m) != cov_func.D:
+            raise ValueError(arg_err_msg)
+        m = tuple(m)
+        if L is not None and (not isinstance(L, Sequence) or len(L) != cov_func.D):
+            raise ValueError(arg_err_msg)
+        elif L is not None:
+            L = tuple(L)
+
+        if L is None and c < 1.2:
+            warnings.warn(
+                "Most applications will require a `c >= 1.2` for accuracy at the boundaries of the "
+                "domain."
+            )
+
+        self.drop_first = drop_first
+        self.m = m
+        self.L = L
+        self.c = c
+        self.D = cov_func.D
+
+        super().__init__(mean_func=mean_func, cov_func=cov_func)
+
+    def __add__(self, other):
+        raise NotImplementedError("Additive HSGPs aren't supported ")
+
+    def _set_boundary(self, X):
+        """Make L from X and c if L is not passed in."""
+        if self.L is None:
+            # Define new L based on c and X range
+            La = at.abs(at.min(X, axis=0))
+            Lb = at.abs(at.max(X, axis=0))
+            self.L = self.c * at.max(at.stack((La, Lb)), axis=0)
+        else:
+            self.L = at.as_tensor_variable(self.L)
+
+    @staticmethod
+    def _eigendecomposition(X, L, m, D):
+        """Construct the eigenvalues and eigenfunctions of the Laplace operator."""
+        m_star = at.prod(m)
+        S = np.meshgrid(*[np.arange(1, 1 + m[d]) for d in range(D)])
+        S = np.vstack([s.flatten() for s in S]).T
+        eigvals = at.square((np.pi * S) / (2 * L))
+        phi = at.ones((X.shape[0], m_star))
+        for d in range(D):
+            c = 1.0 / np.sqrt(L[d])
+            phi *= c * at.sin(at.sqrt(eigvals[:, d]) * (at.tile(X[:, d][:, None], m_star) + L[d]))
+        omega = at.sqrt(eigvals)
+        return omega, phi, m_star
+
+    def approx_K(self, X):
+        """A helper function which gives the approximate kernel or covariance matrix K. This can be
+        helpful when trying to see how well an approximation may work.
+        """
+        X, _ = self.cov_func._slice(X)
+        self._set_boundary(X)
+        omega, phi, _ = self._eigendecomposition(X, self.L, self.m, self.cov_func.D)
+        psd = self.cov_func.psd(omega)
+        return at.dot(phi * psd, at.transpose(phi))
+
+    def prior(self, name, X, dims=None):
+        R"""
+        Returns the (approximate) GP prior distribution evaluated over the input locations `X`.
+
+        Parameters
+        ----------
+        name: string
+            Name of the random variable
+        X: array-like
+            Function input values.
+        dims: None
+            Dimension name for the GP random variable.
+        """
+
+        X, _ = self.cov_func._slice(X)
+        self._set_boundary(X)
+        omega, phi, m_star = self._eigendecomposition(X, self.L, self.m, self.D)
+        psd = self.cov_func.psd(omega)
+
+        if self.drop_first:
```

Comment:
Consider simplification to remove mostly redundant `if` statement: set `i = int(self.drop_first==True)` and then use `i` instead of 1 in the subtraction and indices.

Reply:
much better...